tools/nimgrep.nim

```diff
@@ -494,28 +710,342 @@ proc styleInsensitive(s: string): string =
+  if walkOpt.extensions.len != 0:
+    var matched = false
+    for x in walkOpt.extensions:
+      if os.cmpPaths(x, ex) == 0:
+        matched = true
+        break
+    if not matched: return false
+  for x in walkOpt.skipExtensions:
+    if os.cmpPaths(x, ex) == 0: return false
+  if walkOptC.includeFile.len != 0:
+    var matched = false
+    for pat in walkOptC.includeFile:
+      if filename.match(pat):
+        matched = true
+        break
+    if not matched: return false
+  for pat in walkOptC.excludeFile:
+    if filename.match(pat): return false
+  let dirname = path.parentDir
+  if walkOptC.includeDir.len != 0:
+    var matched = false
+    for pat in walkOptC.includeDir:
+      if dirname.match(pat):
+        matched = true
+        break
+    if not matched: return false
+  result = true
+
+proc hasRightDirectory(path: string, walkOptC: WalkOptComp[Pattern]): bool =
+  let dirname = path.lastPathPart
+  for pat in walkOptC.excludeDir:
+    if dirname.match(pat): return false
+  result = true
+
+iterator walkDirBasic(dir: string, walkOptC: WalkOptComp[Pattern]): string =
+  var dirStack = @[dir]  # stack of directories
+  var timeFiles = newSeq[(times.Time, string)]()
+  while dirStack.len > 0:
+    let d = dirStack.pop()
+    var files = newSeq[string]()
+    var dirs = newSeq[string]()
+    for kind, path in walkDir(d):
       case kind
       of pcFile:
-        if path.hasRightFileName:
-          processFile(pattern, path, counter, errors)
+        if path.hasRightFileName(walkOptC):
+          files.add(path)
       of pcLinkToFile:
-        if optFollow in options and path.hasRightFileName:
-          processFile(pattern, path, counter, errors)
+        if optFollow in options and path.hasRightFileName(walkOptC):
+          files.add(path)
       of pcDir:
-        if optRecursive in options and path.hasRightDirectory:
-          walker(pattern, path, counter, errors)
+        if optRecursive in options and path.hasRightDirectory(walkOptC):
+          dirs.add path
       of pcLinkToDir:
         if optFollow in options and optRecursive in options and
-           path.hasRightDirectory:
-          walker(pattern, path, counter, errors)
-  elif fileExists(dir):
-    processFile(pattern, dir, counter, errors)
+           path.hasRightDirectory(walkOptC):
+          dirs.add path
+    if sortTime:  # sort by time - collect files before yielding
+      for file in files:
+        var time: Time
+        try:
+          time = getLastModificationTime(file)  # can fail for broken symlink
+        except:
+          discard
+        timeFiles.add((time, file))
+    else:  # alphanumeric sort, yield immediately after sorting
+      files.sort()
+      for file in files:
+        yield file
+      dirs.sort(order = SortOrder.Descending)
+    for dir in dirs:
+      dirStack.add(dir)
+  if sortTime:
+    timeFiles.sort(sortTimeOrder)
+    for (_, file) in timeFiles:
+      yield file
+
```

Comment:
performance seems bad, not exactly sure whether:
with -d:danger --gc:arc:
nimgrep -r --filenames . | wc -l
(63723 files)
5.5 seconds

with https://github.com/nim-lang/Nim/pull/15598:
```
import std/[globs,os,sugar]
for ai in glob(".", relative = true, sortCmp = (a,b) => cmp(a.path, b.path)):
  if ai.kind == pcFile: echo ai.path
```
0.22 seconds

echo 'echo 5.5 / 0.22' | nim r -
=> 25 slower

Reply:
You'd better use command `bin/nimgrep -r --filenames '.*' | wc -l`, in my case it improved speed by factor **11x** w.r.t. to matching by '.'

The problem here is that the first positional argument that you pass to nimgrep is actually your pattern (not a directory to search in).
So you are essentially matching any char in a file name and then passing every char one by one to the printing proc.

You may have been confused because it prints every filename on one line. It's because (starting from nimgrep improvements 1.5) nimgrep prints matches in a smart way, displaying all matches in one line. For example, when you match file 'abc' by '.' in old nimgrep 1.4 it will show:
<ins>a</ins>bc
a<ins>b</ins>c
ab<ins>c</ins>
While nimgrep 1.5/1.6 shows:
<ins>abc</ins>
in both cases of matching by '.' and '.*'
But in the first case the chars are passed one-by-one.